File: igris-overture/sdk/go/schlep/escapevector/cache.go
package escapevector

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

// InertialCache manages 72-hour persistent Bayesian state cache
type InertialCache struct {
	cachePath string
	signer    *BayesianSigner
	mu        sync.RWMutex
}

// NewInertialCache creates a new inertial cache
func NewInertialCache(cacheDir string, encryptionKey []byte) (*InertialCache, error) {
	// Ensure cache directory exists
	if err := os.MkdirAll(cacheDir, 0700); err != nil {
		return nil, fmt.Errorf("failed to create cache directory: %w", err)
	}

	signer, err := NewBayesianSigner(encryptionKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create signer: %w", err)
	}

	return &InertialCache{
		cachePath: filepath.Join(cacheDir, "bayesian_state.enc"),
		signer:    signer,
	}, nil
}

// Save encrypts and saves Bayesian state to disk
func (ic *InertialCache) Save(state *BayesianState) error {
	ic.mu.Lock()
	defer ic.mu.Unlock()

	// Encrypt state
	encrypted, err := ic.signer.EncryptState(state)
	if err != nil {
		return fmt.Errorf("failed to encrypt state: %w", err)
	}

	// Serialize encrypted state
	data, err := json.Marshal(encrypted)
	if err != nil {
		return fmt.Errorf("failed to marshal encrypted state: %w", err)
	}

	// Write to disk atomically
	tempPath := ic.cachePath + ".tmp"
	if err := os.WriteFile(tempPath, data, 0600); err != nil {
		return fmt.Errorf("failed to write temp file: %w", err)
	}

	if err := os.Rename(tempPath, ic.cachePath); err != nil {
		os.Remove(tempPath)
		return fmt.Errorf("failed to rename temp file: %w", err)
	}

	return nil
}

// Load decrypts and loads Bayesian state from disk
func (ic *InertialCache) Load() (*BayesianState, error) {
	ic.mu.RLock()
	defer ic.mu.RUnlock()

	// Read encrypted state
	data, err := os.ReadFile(ic.cachePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("no cached state found")
		}
		return nil, fmt.Errorf("failed to read cache file: %w", err)
	}

	// Deserialize
	var encrypted EncryptedBayesianState
	if err := json.Unmarshal(data, &encrypted); err != nil {
		return nil, fmt.Errorf("failed to unmarshal encrypted state: %w", err)
	}

	// Decrypt and verify
	state, err := ic.signer.DecryptState(&encrypted)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt state (may be expired or tampered): %w", err)
	}

	return state, nil
}

// Exists checks if cache file exists
func (ic *InertialCache) Exists() bool {
	ic.mu.RLock()
	defer ic.mu.RUnlock()

	_, err := os.Stat(ic.cachePath)
	return err == nil
}

// Clear removes the cache file
func (ic *InertialCache) Clear() error {
	ic.mu.Lock()
	defer ic.mu.Unlock()

	if err := os.Remove(ic.cachePath); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("failed to clear cache: %w", err)
	}
	return nil
}

File: igris-overture/sdk/javascript/src/escapevector/cache.ts
/**
 * Inertial Cache - 72-hour persistent Bayesian state storage
 */

import * as fs from 'fs/promises';
import * as path from 'path';
import * as os from 'os';
import { BayesianState, BayesianSigner, EncryptedBayesianState } from './bayesian-state';

/**
 * Inertial Cache Manager
 */
export class InertialCache {
  private cachePath: string;
  private signer: BayesianSigner;

  constructor(cacheDir?: string, signer?: BayesianSigner) {
    // Default cache directory
    if (!cacheDir) {
      const homeDir = os.homedir();
      cacheDir = path.join(homeDir, '.config', 'schlep');
    }

    this.cachePath = path.join(cacheDir, 'bayesian_state.enc');
    this.signer = signer || new BayesianSigner();
  }

  /**
   * Initialize with encryption key
   */
  async init(keyBytes: Uint8Array): Promise<void> {
    await this.signer.init(keyBytes);

    // Ensure cache directory exists
    const dir = path.dirname(this.cachePath);
    await fs.mkdir(dir, { recursive: true, mode: 0o700 });
  }

  /**
   * Save encrypted Bayesian state
   */
  async save(state: BayesianState): Promise<void> {
    // Encrypt state
    const encrypted = await this.signer.encryptState(state);

    // Serialize encrypted state
    const data = JSON.stringify({
      ciphertext: Buffer.from(encrypted.ciphertext).toString('base64'),
      iv: Buffer.from(encrypted.iv).toString('base64'),
      hmac: Buffer.from(encrypted.hmac).toString('base64'),
      version: encrypted.version,
      timestamp: encrypted.timestamp,
      expiresAt: encrypted.expiresAt,
    });

    // Write atomically
    const tempPath = this.cachePath + '.tmp';
    await fs.writeFile(tempPath, data, { mode: 0o600 });
    await fs.rename(tempPath, this.cachePath);
  }

  /**
   * Load and decrypt Bayesian state
   */
  async load(): Promise<BayesianState> {
    try {
      const data = await fs.readFile(this.cachePath, 'utf-8');
      const parsed = JSON.parse(data);

      const encrypted: EncryptedBayesianState = {
        ciphertext: Buffer.from(parsed.ciphertext, 'base64').buffer,
        iv: Buffer.from(parsed.iv, 'base64').buffer,
        hmac: Buffer.from(parsed.hmac, 'base64').buffer,
        version: parsed.version,
        timestamp: parsed.timestamp,
        expiresAt: parsed.expiresAt,
      };

      return await this.signer.decryptState(encrypted);
    } catch (error) {
      if ((error as any).code === 'ENOENT') {
        throw new Error('No cached state found');
      }
      throw new Error(
        `Failed to load cache (may be expired or tampered): ${(error as Error).message}`
      );
    }
  }

  /**
   * Check if cache exists
   */
  async exists(): Promise<boolean> {
    try {
      await fs.access(this.cachePath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Clear cache
   */
  async clear(): Promise<void> {
    try {
      await fs.unlink(this.cachePath);
    } catch (error) {
      if ((error as any).code !== 'ENOENT') {
        throw new Error(`Failed to clear cache: ${(error as Error).message}`);
      }
    }
  }
}
